### PR TITLE
Ros2 lights tests

### PR DESCRIPTION
**Pull request by pawelirh**
### Description

Restructured unit tests to be independent of hardware

- Add interfaces for mocked classes
- Restructure unit tests to avoid using ROS communication and hardware

### Requirements

- [X] Code style guidelines followed
- [X] Documentation updated

### Tests 🧪

- [X] Robot
- [ ] Container
- [ ] Simulation

## Summary by CodeRabbit

- **New Features**
	- Enhanced testing capabilities with the integration of Google Mock and ROS testing frameworks.
	- Introduced new interfaces for SPI device handling and LED panel control, improving modularity and testability.
	- Updated `LightsDriverNode` to support shared pointers for LED channel management.

- **Bug Fixes**
	- Corrected typographical errors in variable names for improved readability.

- **Tests**
	- Added new test cases for `APA102Wrapper` and `LightsDriverNode` to validate behavior under various scenarios.
	- Refined existing tests to improve coverage and clarity, ensuring robust error handling.

### CODE REVIEW

File: panther_lights/test/test_lights_driver_node.cpp [hunk 1]
```diff
@@ -14,191 +14,287 @@
...
+  std::unique_ptr<DriverNodeWrapper> lights_driver_node_;
+};
+
+TestLightsDriverNode::TestLightsDriverNode()
+{
+  channel_1_ = std::make_shared<MockAPA102::NiceMock>();
+  channel_2_ = std::make_shared<MockAPA102::NiceMock>();
+
+  lights_driver_node_ = std::make_unique<DriverNodeWrapper>(channel_1_, channel_2_);
+}
```
[the remark this answers is not in the document]
Comment (pawelirh) [reply panther_lights/test/test_lights_driver_node.cpp:116]: That's fine, it's only a TestFixture class ;)

File: panther_lights/test/test_apa102.cpp [hunk 2]
```diff
@@ -31,52 +50,121 @@ class APA102Wrapper : public panther_lights::apa102::APA102
...
 
-TEST_F(TestAPA102, PortsAvailable)
+TestAPA102::TestAPA102()
+{
+  spi_device_ = std::make_shared<MockSPIDevice::NiceMock>();
+
+  apa102_ = std::make_unique<APA102Wrapper>(spi_device_, kMockDeviceName);
+}
+
+TEST_F(TestAPA102, TestInitialization) { EXPECT_TRUE(spi_device_ != nullptr); }
```
Comment (KmakD) [inline panther_lights/test/test_apa102.cpp]: Not sure what we are testing here

File: panther_lights/test/test_apa102.cpp [hunk 3]
```diff
@@ -31,52 +50,121 @@ class APA102Wrapper : public panther_lights::apa102::APA102
...
+
+TEST(TestInitialization, InvalidDevices)
+{
+  auto spi_device = std::make_shared<MockSPIDevice::NiceMock>();
+
+  // Return -1 to simulate failed device opening
+  ON_CALL(*spi_device, Open(kMockDeviceName)).WillByDefault(testing::Return(-1));
+
+  EXPECT_THROW(std::make_unique<APA102Wrapper>(spi_device, kMockDeviceName);
+               , std::ios_base::failure);
```
Comment (KmakD) [inline panther_lights/test/test_apa102.cpp]:
```suggestion
  EXPECT_THROW(APA102Wrapper(spi_device, kMockDeviceName)
  , std::ios_base::failure);
```
Don't think we need pointer here. Also, is this semicolon necessary here?
Applies to other initialization tests.

File: panther_lights/test/test_lights_driver_node.cpp [hunk 4]
```diff
@@ -14,191 +14,287 @@
...
+{
+  auto msg = CreateValidImageMsg();
+
+  auto future = CreateSetBoolSrvFuture(true, true);
+  lights_driver_node_->ToggleLEDControlCB(std::move(future));
+
+  EXPECT_CALL(*channel_1_, SetPanel(testing::_)).Times(1);
+
+  lights_driver_node_->FrameCB(msg, channel_1_, msg->header.stamp, "channel_1");
+}
```
Comment (KmakD) [inline panther_lights/test/test_lights_driver_node.cpp]: Can we also add check for channel_2 for successful call?

**Review by KmakD — CHANGES_REQUESTED** [review]

File: panther_lights/test/test_apa102.cpp (hunk 2 above)
Comment (pawelirh) [reply panther_lights/test/test_apa102.cpp]: You're right, this may be removed 👍🏼

**Review by KmakD — APPROVED** [review]

**Pull request merged**